returnn/sprint/control.py

```diff
@@ -403,11 +403,41 @@ def _get_loss_and_error_signal_via_sprint_callback(self, seg_name, orthography,
     return loss, error_signal
 
   def _send(self, data):
-    Pickler(self.pipe_c2p).dump(data)
+    #Pickler(self.pipe_c2p).dump(data)
+    import struct
+    stream = BytesIO()
+    Pickler(stream).dump(data)
+    raw_data = stream.getvalue()
+    assert len(raw_data) > 0
+    self.pipe_c2p.write(struct.pack("<i", len(raw_data)))
+    self.pipe_c2p.write(raw_data)
     self.pipe_c2p.flush()
 
   def _read(self):
-    return Unpickler(self.pipe_p2c).load()
+    #return Unpickler(self.pipe_p2c).load()
+    import struct
+    size_raw = self.pipe_p2c.read(4)
+    if len(size_raw) < 4:
+      raise EOFError
+    size, = struct.unpack("<i", size_raw)
+    assert size > 0, "%s: We expect to get some non-empty package. Invalid Python mod in Sprint?" % (self,)
+    stream = BytesIO()
+    read_size = 0
+    while read_size < size:
+      data_raw = self.pipe_p2c.read(size - read_size)
+      if len(data_raw) == 0:
+        raise EOFError("%s: expected to read %i bytes but got EOF after %i bytes" % (self, size, read_size))
+      read_size += len(data_raw)
+      stream.write(data_raw)
+    stream.seek(0)
+    try:
+      if PY3:
+        data = Unpickler(stream, encoding="bytes").load()
```

Comment:
It seems a lot of this code was copied from `ExternSprintDataset._read_next_raw`. But why did you remove the comments from there? The comments should be left in.

Reply:
> It seems a lot of this code was copied from `ExternSprintDataset._read_next_raw`. But why did you remove the comments from there? The comments should be left in.

I am not familiar with this and I just applied Wei's changes as they are.

Reply:
> I am not familiar with this and I just applied Wei's changes as they are.

That's not really relevant here. What matters is that the code changes are done in a clean and correct way.